Libraries/Lists/VirtualizedList.js

```diff
@@ -734,6 +750,163 @@ class VirtualizedList extends React.PureComponent<Props, State> {
     this.state = initialState;
   }
 
+  static _prepareState(
+    props: Props,
+    viewportWindow: {first: number, last: number},
+  ): State {
+    const itemCount = props.getItemCount(props.data);
+
+    invariant(
+      viewportWindow.first >= 0 &&
+        viewportWindow.last >= viewportWindow.first - 1 &&
+        viewportWindow.last < itemCount,
+      `Invalid viewport window "${JSON.stringify(
+        viewportWindow,
+      )}" was passed to VirtualizedList._prepareState`,
+    );
+
+    const renderMask = new CellRenderMask(itemCount);
+
+    if (itemCount > 0) {
+      renderMask.addCells(viewportWindow);
+
+      const scrollIndex = props.initialScrollIndex || 0;
+      if (scrollIndex === 0) {
+        const initialRegion = VirtualizedList._initialRenderRegion(props);
+        renderMask.addCells(initialRegion);
+      }
+
+      const stickyIndicesSet = new Set(props.stickyHeaderIndices);
+      VirtualizedList._ensureStickyHeadersBefore(
+        props,
+        stickyIndicesSet,
+        renderMask,
+        viewportWindow.first,
+      );
+    }
+
+    return {renderMask, viewportWindow};
+  }
+
+  static _initialRenderRegion(props: Props): {first: number, last: number} {
+    const itemCount = props.getItemCount(props.data);
+    const scrollIndex = props.initialScrollIndex || 0;
+
+    return {
+      first: scrollIndex,
+      last:
+        Math.min(
+          itemCount,
+          scrollIndex + initialNumToRenderOrDefault(props.initialNumToRender),
+        ) - 1,
```

Comment:
This allows something odd like `[0, -1]`, but that is previous behavior for how we represent a zero-item range.